File: udp_server.go
/*
 * UDP Server in Go
 * Listens on port 8083
 * Prints "Go server received: <message>" when a message is received
 * Handles Ctrl+C gracefully
 *
 * How to compile:
 *     go build -o udp_server_go udp_server.go
 *
 * How to run:
 *     ./udp_server_go
 *
 * Press Ctrl+C to stop the server.
 */

package main

import (
	"fmt"
	"net"
	"os"
	"os/signal"
	"syscall"
	"time"
)

func main() {
	// Create a UDP address to listen on
	addr := net.UDPAddr{
		Port: 8083,
		IP:   net.ParseIP("127.0.0.1"),
	}

	// Listen for incoming UDP packets
	conn, err := net.ListenUDP("udp", &addr)
	if err != nil {
		fmt.Println("Error starting server:", err)
		return
	}
	defer conn.Close()
	fmt.Println("Go UDP server listening on port 8083")

	// Channel to handle OS signals
	sigs := make(chan os.Signal, 1)
	signal.Notify(sigs, syscall.SIGINT, syscall.SIGTERM)

	// Channel to signal the server to stop
	done := make(chan bool, 1)

	// Goroutine to handle incoming messages
	go func() {
		buf := make([]byte, 1024)
		for {
			select {
			case <-done:
				return
			default:
				conn.SetReadDeadline(time.Now().Add(1 * time.Second))
				n, _, err := conn.ReadFromUDP(buf)
				if err != nil {
					if netErr, ok := err.(net.Error); ok && netErr.Timeout() {
						continue
					}
					fmt.Println("Error reading:", err)
					continue
				}
				message := string(buf[:n])
				fmt.Printf("Go server received: %s\n", message)
			}
		}
	}()

	// Wait for Ctrl+C (SIGINT)
	<-sigs
	fmt.Println("\nServer terminated gracefully")
	close(done)
}

File: udp_server.cpp
/*
 * UDP Server in C++
 * Listens on port 8081
 * Prints "C++ server received: <message>" when a message is received
 * Handles Ctrl+C gracefully
 *
 * How to compile:
 *     g++ -o udp_server_cpp udp_server.cpp
 *
 * How to run:
 *     ./udp_server_cpp
 *
 * Press Ctrl+C to stop the server.
 */

#include <arpa/inet.h> // For sockaddr_in and inet_ntoa()
#include <csignal>
#include <cstdlib>
#include <cstring>
#include <iostream>
#include <sys/socket.h> // For socket functions
#include <unistd.h>     // For close()

int sockfd; // Socket file descriptor

void handle_sigint(int sig) {
  // Close the socket and exit
  close(sockfd);
  std::cout << "\nServer terminated gracefully" << std::endl;
  exit(0);
}

int main() {
  struct sockaddr_in server_addr, client_addr;
  char buffer[1024];
  socklen_t addr_len = sizeof(client_addr);

  // Set up the signal handler for Ctrl+C (SIGINT)
  signal(SIGINT, handle_sigint);

  // Create a UDP socket
  if ((sockfd = socket(AF_INET, SOCK_DGRAM, 0)) < 0) {
    perror("Socket creation failed");
    exit(EXIT_FAILURE);
  }

  // Zero out the server address structure
  memset(&server_addr, 0, sizeof(server_addr));

  // Set up the server address
  server_addr.sin_family = AF_INET;         // IPv4
  server_addr.sin_addr.s_addr = INADDR_ANY; // Bind to all interfaces
  server_addr.sin_port = htons(8081);       // Port 8081

  // Bind the socket to the server address
  if (bind(sockfd, (const struct sockaddr *)&server_addr, sizeof(server_addr)) <
      0) {
    perror("Bind failed");
    close(sockfd);
    exit(EXIT_FAILURE);
  }

  std::cout << "C++ UDP server listening on port 8081" << std::endl;

  while (true) {
    // Receive a message from a client
    ssize_t n = recvfrom(sockfd, buffer, sizeof(buffer) - 1, 0,
                         (struct sockaddr *)&client_addr, &addr_len);
    if (n < 0) {
      perror("Receive failed");
      continue;
    }

    buffer[n] = '\0'; // Null-terminate the received data

    // Print the received message
    std::cout << "C++ server received: " << buffer << std::endl;
  }

  // Close the socket (unreachable code in this example)
  close(sockfd);

  return 0;
}
